src/cultionet/data/datasets.py

```diff
@@ -419,10 +424,46 @@ def split_train_val(
         Returns:
             train dataset, validation dataset
         """
-        n_train = int(len(self) * (1.0 - val_frac))
+        id_column = 'common_id'
         self.shuffle_items()
-        train_ds = self[:n_train]
-        val_ds = self[n_train:]
+        if spatial_overlap_allowed:
+            n_train = int(len(self) * (1.0 - val_frac))
+            train_ds = self[:n_train]
+            val_ds = self[n_train:]
+        else:
+            self.create_spatial_index()
+            self.dataset_df[id_column] = (
+                self.dataset_df.grid_id
+                .str.split('_', expand=True).loc[:, 0]
+            )
+            unique_ids = self.dataset_df.common_id.unique()
+            if spatial_balance:
+                df_unique_locations = gpd.GeoDataFrame(
+                    pd.Series(unique_ids)
+                    .to_frame(name=id_column)
+                    .merge(self.dataset_df, on=id_column)
+                    .drop_duplicates(id_column)
+                    .drop(columns=['grid_id'])
+                ).to_crs('EPSG:8858')
+
+                qt = QuadTree(df_unique_locations, force_square=False)
+                qt.split_recursive(max_samples=1)
+                n_val = int(val_frac * len(df_unique_locations.index))
+                df_val_sample = qt.sample(n=n_val)
```

Comment:
What does this do? Something regarding the spatial distribution of the validation set, but it's not totally clear to me.